# Support Vector Regression (SVR) Project  

## Tesla Model Y Price Prediction | End-to-End ML Project  

This project focuses on predicting the price of Tesla Model Y cars using **Support Vector Regression (SVR)**.

It covers the full machine learning workflow, including data preprocessing, exploratory data analysis (EDA), feature selection, and model training.

---

## Dataset  

The dataset used in this project is available on Kaggle: [Dataset Link](https://www.kaggle.com/datasets/aiwithcagri/multiple-linear-regression-tesla-model-y-price)
  

---

## Objective  

The main objective is to build a regression model that can accurately estimate car prices based on key features such as:
- year
- km  
- color 

---

## Workflow  

The project follows these steps:

1. **Data Cleaning**

2. **Exploratory Data Analysis (EDA)**

3. **Feature Engineering** 

4. **Model Building**

5. **Model Evaluation**

---

## Tools & Technologies  

- Python  
- Pandas  
- NumPy  
- Scikit-learn  

---

## Key Learnings  

- Understanding how multiple features affect predictions  
- Importance of proper data preprocessing  
- Handling categorical variables correctly  
- Building and interpreting regression models  

---

## Next Steps  

- Apply Decision Tree 
- Compare with other regression models (Multiple Linear Regression, Decision Tree, Random Forest)  
- Improve model performance with feature engineering  

# **Importing The Libaries**

In [92]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data Cleaning Check Template**
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:


In [93]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [94]:
dataset = pd.read_csv("/kaggle/input/datasets/aiwithcagri/multiple-linear-regression-tesla-model-y-price/teslay.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255 entries, 0 to 254
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   year    255 non-null    int64  
 1   km      255 non-null    float64
 2   color   255 non-null    object 
 3   price   255 non-null    int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 8.1+ KB
None

SHAPE OF DATA
Rows: 255, Columns: 4

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 4

DATA TYPES
year       int64
km       float64
color     object
price      int64
dtype: object

NUMERICAL SUMMARY
              year             km         price
count   255.000000     255.000000  2.550000e+02
mean   2024.207843   30384.172549  2.247061e+06
std       0.406562   15563.810709  1.517529e+05
min    2024.000000    6001.000000  1.825000e+06
25%    2024.000000   19880.000000  2.160000e+06
50%    2024.000000   29000.000000  2.239750e+06
75%    2024.000000 

# **Understaning the outliers** 

### km: 6 outliers

In [95]:
# Calculate fences for Price
Q1 = dataset['km'].quantile(0.25)
Q3 = dataset['km'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
km_outliers = dataset[(dataset['km'] < lower_fence) | (dataset['km'] > upper_fence)]

# Display the rows
print(km_outliers)

     year        km  color    price
0    2025   69000.0  White  2095000
23   2024   73000.0   Blue  2140000
96   2024   75000.0  White  2010000
145  2024  110000.0  White  2040000
210  2024   96000.0  Black  2049999
229  2024   68200.0  White  2100000


In this dataset, the KM (Mileage) column contains a few rows with exceptionally high values, such as 110,000 km for a 2024 model. This creates three major issues for our Machine Learning model:

1. Statistical Noise: Because the vast majority of our cars have low mileage (around 20,000–30,000 km), these extreme points act as "noise." They pull the mathematical average (mean) away from the typical car, making the data less representative of the standard market.


2. Logic Conflict: A car driven over 100,000 km in a single year is likely a commercial vehicle or a taxi. These cars have significantly more wear-and-tear than private vehicles. Including them confuses the model because their relationship between year and Price is completely different from the rest of the data.


3. Model Skew: When building a Linear Regression model, these distant points can "tilt" the prediction line. The model will try so hard to reach these outliers that it will lose accuracy when predicting prices for "normal" low-mileage cars.

Conclusion

To ensure the model remains accurate and focused on the standard consumer market, these high-mileage outliers must be removed from the dataset.

# **Cleaning the data** 

### Remove km outliers (6) & remove the duplication (4)

6 + 4 = 10 Rows

255 - 10 = 245 Rows

shape = 245 * 4 

In [96]:
# Remove rows that exist in km_outliers from the original dataset
dataset_clean = dataset.drop(km_outliers.index).copy()

# Step 2: Remove duplicates FROM the cleaned dataset
dataset_clean = dataset_clean.drop_duplicates().copy()

# Final result
print(dataset_clean)

     year       km  color    price
1    2024  28500.0  Black  2280000
2    2024  50000.0  White  2090000
3    2024  42000.0  White  2180000
4    2024  15000.0  White  2300000
5    2024  36326.0  White  2155000
..    ...      ...    ...      ...
250  2025  24000.0  Black  2390000
251  2024  29000.0  White  2185000
252  2025  19800.0  White  2305000
253  2024  16000.0  White  2250000
254  2024  34000.0  White  2310000

[245 rows x 4 columns]


# **Select Independent Variable (X) and Dependent Variable (y)**

📌 Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [97]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset_clean.iloc[:, :-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset_clean.iloc[:, -1].values

In [98]:
# reshape y before to apply feature scaling
y = y.reshape(len(y), 1)

# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

📌 Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [99]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False, drop='first'),[2])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))

# **Splitting the dataset into the Train set and Test set**

In [100]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train , y_test = train_test_split(X, y , test_size = 0.2, random_state = 0)

In [101]:
print (X_test)

[[0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 25000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 22700.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 10124.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 33000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 66000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 41300.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 7650.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 6001.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 37000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 44000.0]
 [0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 31000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2025 19200.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 6001.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2025 19200.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 52531.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2025 41200.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 29000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 25000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2025 41500.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 21000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2025 3700

# **Feature Scaling**

📌 Note: I explained Feature Scaling earlier.

If you want to understand when and why to use feature scaling, you can check it here: [Go to  When is Feature Scaling Necessary in Regression Models? Section](https://github.com/Hazem1695/ml-concept-briefs)

In [102]:
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X_train[:, [8, 9]] = sc_X.fit_transform(X_train[:,[8, 9]])
y_train = sc_y.fit_transform(y_train)
X_test[:, [8, 9]] = sc_X.transform(X_test[:, [8, 9]])

In [103]:
print(X_test)

[[0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182
  -0.32628800933730856]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 -0.5011413295737486]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 -1.4572088753535177]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 0.2818974523546566]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 2.790662481834013]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 0.9128898688600704]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 -1.645290229381758]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 -0.5063696835417182 -1.7706524576730143]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 0.5859901832006391]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 1.1181524621811088]
 [0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 -0.5063696835417182 0.1298510869316653]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.974841765813265 -0.7672224690639833]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 -1.7706524576730143]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.

In [104]:
print(f"Rows: {X_test.shape[0]}, Columns: {X_test.shape[1]}")

Rows: 49, Columns: 10


# **Building Support Vector Regression (SVR) Model**

## Training the Building Support Vector Regression (SVR) model on the Training set

In [105]:
from sklearn.svm import SVR
regressor = SVR(kernel = 'rbf')
regressor.fit(X_train, y_train.ravel())

SVR()

# **Preticting the Test set results**

In [106]:
y_pred = sc_y.inverse_transform(regressor.predict(X_test).reshape(-1,1))
np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[2216768.77 2185000.  ]
 [2230700.79 2250000.  ]
 [2355479.86 2275000.  ]
 [2184864.95 2249990.  ]
 [2140416.39 2089000.  ]
 [2153454.7  2199000.  ]
 [2381789.58 2350000.  ]
 [2442175.63 2425000.  ]
 [2170921.46 2100000.  ]
 [2141534.97 2050000.  ]
 [2287683.97 2320000.  ]
 [2321106.97 2350000.  ]
 [2397461.45 2490000.  ]
 [2326967.72 2365000.  ]
 [2152942.76 2125000.  ]
 [2196028.74 2400000.  ]
 [2198741.45 2265555.  ]
 [2280678.14 2340000.  ]
 [2216780.91 2180000.  ]
 [2300114.33 2350000.  ]
 [2218009.43 2165000.  ]
 [2224726.11 2189000.  ]
 [2218414.27 2185000.  ]
 [2206657.92 2215000.  ]
 [2235338.12 2199000.  ]
 [2195784.06 2240000.  ]
 [2255533.76 2199000.  ]
 [2188197.73 2075000.  ]
 [2397452.53 2500000.  ]
 [2355650.62 2285000.  ]
 [2167107.99 1990000.  ]
 [2206945.51 2130000.  ]
 [2323998.66 2300000.  ]
 [2196145.96 2280000.  ]
 [2228024.66 2165000.  ]
 [2167107.99 1950000.  ]
 [2356854.97 2490000.  ]
 [2163135.26 2269900.  ]
 [2108845.62 2195000.  ]
 [2217019.57 2240000.  ]


# **Single Prediction**

In [110]:
# 1. New input
X_new = [[2024, 30000, 'White']]

# 2. Apply same encoding used in training
X_new = ct.transform(X_new)

# 3. Apply SAME scaler used in training
X_new[:, -2:] = sc_X.transform(X_new[:, -2:])


y_pred_single = sc_y.inverse_transform(regressor.predict(X_new).reshape(-1,1))


print(y_pred_single)

[[2195066.15]]


In [109]:
print(X_new)

[[0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 -0.5063696835417182 0.05382790422016965]]


# **Evaluating the Model Performance**

In [107]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.5324211595666584

# **Important Insight**
1. The SVR model achieved its best performance (R² = 0.53) when feature scaling was applied selectively to high-magnitude numerical features (Year, KM).

2. Applying scaling to the entire feature set, including dummy variables, significantly degraded performance (R² = 0.23).

3. This indicates that uniform preprocessing is not always optimal, even for models that are known to be sensitive to feature scaling.

4. One Hot encoded features inherently lie within a bounded range (0 and 1), and applying feature scaling to them can distort their representation and weaken their contribution to the model.

5. The results emphasize that feature-type awareness is critical when designing preprocessing pipelines.

---
# **Conclusion**

The SVR model demonstrated that preprocessing decisions have a significant impact on model performance. While feature scaling is generally required for SVR, applying it indiscriminately across all features led to a substantial drop in performance.

The best results were achieved by scaling only the high-magnitude numerical features (Year and KM), resulting in an R² score of 0.53, compared to 0.23 when all features were scaled.

This suggests that effective preprocessing requires a targeted approach that considers the nature of each feature rather than applying uniform transformations.

These findings reinforce the importance of experimental validation in machine learning workflows, where assumptions about preprocessing techniques must be tested rather than blindly applied.

---
# **Model Comparison**
| Model                | Configuration                          | R² Score |
|---------------------|----------------------------------------|---------|
| Linear Regression   | No Scaling                             | 0.42    |
| Linear Regression   | With Scaling                           | 0.41    |
| Polynomial Regression | With Scaling                         | 0.48    |
| SVR                | Scale All Features                      | 0.23    |
| SVR                | Scale Numerical Features (Year, KM)     | 0.53    |